Example app with external service registration (#138)
grafana/grafana-plugin-examples: examples/app-with-on-behalf-of-auth/README.md

Reviewer: I would replace the `service account` with `plugin` almost everywhere, because it is just a detail and a "special" type of service account which will be hidden from the normal service accounts' list.

# Grafana app plugin with OAuth2 integration

This plugin is an example of how to integrate OAuth2 authentication into a Grafana plugin.

> **Note:** This plugin requires Grafana 10.1 or later.

Reviewer: ```suggestion
**Note:** This plugin requires Grafana 10.1 or later and the `externalServiceAuth` feature toggle must be enabled. This is an experimental feature.
```

## How to use

> This app allows you to do requests to the Grafana API using either a service account created with the plugin or on behalf of a user (specifying the user ID). The plugin will then use the access token to do requests to the Grafana API.

Reviewer: ```suggestion
This app allows you to do requests to the Grafana API as the plugin or on behalf of a user (by specifying the user ID). The plugin will then use the access token to do requests to the Grafana API.
```

![screenshot](./src/img/screenshot-showcase.png)

## Authentication flow

> The plugin uses the [OAuth2 Authorization Code Flow](https://oauth.net/2/grant-types/authorization-code/) to authenticate users and obtain an access token that can be used to authorize requests against the Grafana API. To enable it, add the section below to your `plugin.json` file.

Reviewer: ```suggestion
The plugin uses the [JWT Bearer Assertion OAuth2 Extension
](https://datatracker.ietf.org/doc/html/rfc7523) to authenticate users and obtain an access token that can be used to authorize requests against the Grafana API. To enable it, add the section below to your `plugin.json` file.
```

```json
  "externalServiceRegistration": {
    "self": {
      "permissions": [{ "action": "dashboards:create", "scope": "folders:uid:general" }]
    },
    "impersonation": {
      "permissions": [
        { "action": "dashboards:create", "scope": "folders:*" },
        { "action": "dashboards:read", "scope": "dashboards:*" },
        { "action": "dashboards:read", "scope": "folders:*" },
        { "action": "folders:read", "scope": "folders:*" }
      ]
    }
  }
```

> The `self` section defines the permissions that the service account will have. The `impersonation` section defines the permissions that the service will have when impersonating a user. Note that for this to work, the user must have permissions as well to perform those actions. See the Grafana documentation about [access control](https://grafana.com/docs/grafana/latest/administration/roles-and-permissions/access-control/) for more information.

Reviewer: ```suggestion
The `self` section defines the set of permissions granted to the plugin. The `impersonation` section defines the set of permissions that are granted to the impersonated user when the plugin makes requests on behalf of a user. Note that for this to work, the user must have permissions as well to perform those actions. See the Grafana documentation about [access control](https://grafana.com/docs/grafana/latest/administration/roles-and-permissions/access-control/) for more information.
```

## Service registration and token retrieval

Once a plugin is registered with an `externalServiceRegistration` section, Grafana will automatically create a service account for it. After that, to use it, there is a function exposed by the Grafana SDK that can be used to retrieve the access token for the service account. This function relies on environment variables that are set with the necessary credentials:

```go
	app.tokenRetriever, err = oauthtokenretriever.New()
	if err != nil {
		return nil, err
	}
```

> Once the token retriever is initialized, it can be used to retrieve access tokens, either for the service account or for a user impersonated by the service account:

Reviewer: ```suggestion
Once the token retriever is initialized, it can be used to retrieve access tokens, either for the plugin or for a user impersonated by the plugin:
```
